curvefs/src/client/fuse_client.cpp

```diff
@@ -561,5 +594,140 @@ CURVEFS_ERROR FuseClient::FuseOpRelease(fuse_req_t req, fuse_ino_t ino,
+      inodeManager_(inodeManager),
+      mdsClient_(mdsClient) {}
+
+CURVEFS_ERROR RenameOperator::GetTxId() {
+    auto rc = dentryManager_->GetTxId(
+        fsId_, parentId_, &srcPartitionId_, &srcTxId_);
+    if (rc != CURVEFS_ERROR::OK) {
+        LOG(ERROR) << "GetTxId failed, retCode = " << rc
+                   << ", fsId = " << fsId_ << ", inodeId = " << parentId_;
+        return rc;
+    }
+
+    rc = dentryManager_->GetTxId(
+        fsId_, newParentId_, &dstPartitionId_, &dstTxId_);
+    if (rc != CURVEFS_ERROR::OK) {
+        LOG(ERROR) << "GetTxId failed, retCode = " << rc
+                   << ", fsId = " << fsId_ << ", inodeId = " << newParentId_;
+    }
+
+    return rc;
+}
+
+// TODO(Wine93): we should improve the check for whether a directory is empty
+CURVEFS_ERROR RenameOperator::CheckOverwrite() {
+    if (dstDentry_.flag() & DentryFlag::TYPE_FILE_FLAG) {
+        return CURVEFS_ERROR::OK;
+    }
+
+    std::list<Dentry> dentrys;
+    auto rc = dentryManager_->ListDentry(dstDentry_.inodeid(), &dentrys, 1);
+    if (rc == CURVEFS_ERROR::OK && !dentrys.empty()) {
+        LOG(ERROR) << "The directory is not empty"
+                   << ", dentry = (" << dstDentry_.ShortDebugString() << ")";
+        rc = CURVEFS_ERROR::NOTEMPTY;
+    }
+
+    return rc;
+}
+
+CURVEFS_ERROR RenameOperator::Precheck() {
+    auto rc = dentryManager_->GetDentry(parentId_, name_, &srcDentry_);
+    if (rc != CURVEFS_ERROR::OK) {
+        LOG(ERROR) << "Failed to get dentry, retCode = " << rc;
+        return rc;
+    }
+
+    rc = dentryManager_->GetDentry(newParentId_, newname_, &dstDentry_);
+    if (rc == CURVEFS_ERROR::NOTEXIST) {
+        return CURVEFS_ERROR::OK;
+    } else if (rc == CURVEFS_ERROR::OK) {
+        oldInodeId_ = dstDentry_.inodeid();
+        return CheckOverwrite();
+    }
+
+    LOG(ERROR) << "Failed to get dentry, retCode = " << rc;
+    return rc;
+}
+
+CURVEFS_ERROR RenameOperator::PrepareTx() {
+    dentry_ = Dentry(srcDentry_);
+    dentry_.set_txid(srcTxId_ + 1);
+    dentry_.set_flag(DentryFlag::DELETE_MARK_FLAG);
+
+    newDentry_ = Dentry(srcDentry_);
+    newDentry_.set_parentinodeid(newParentId_);
+    newDentry_.set_name(newname_);
+    newDentry_.set_txid(dstTxId_ + 1);
+
+    VLOG(1) << "Prepare rename tx: "
+            << "src dentry = (" << dentry_.ShortDebugString() << ")"
+            << ", dst dentry = (" << newDentry_.ShortDebugString() << ")";
+
+    auto rc = dentryManager_->Rename(dentry_, newDentry_);
+    if (rc != CURVEFS_ERROR::OK) {
+        LOG(ERROR) << "Rename failed, retCode = " << rc;
+    }
+    return rc;
+}
+
+CURVEFS_ERROR RenameOperator::CommitTx() {
+    std::vector<PartitionTxId> txIds;
+
+    PartitionTxId partitionTxId;
+    partitionTxId.set_partitionid(srcPartitionId_);
+    partitionTxId.set_txid(srcTxId_ + 1);
+    txIds.push_back(partitionTxId);
+
+    if (srcPartitionId_ != dstPartitionId_) {
+        partitionTxId.set_partitionid(dstPartitionId_);
```

Comment:
Don't need this judgment, following is ok whether dstPartitionId_ and srcPartitionId_ are equal or not

PartitionTxId partitionTxId;
partitionTxId.set_partitionid(dstPartitionId_);
partitionTxId.set_txid(dstTxId_ + 1);
txIds.push_back(partitionTxId);

Reply:
> Don't need this judgment, following is ok whether dstPartitionId_ and srcPartitionId_ are equal or not
> 
> PartitionTxId partitionTxId;
> partitionTxId.set_partitionid(dstPartitionId_);
> partitionTxId.set_txid(dstTxId_ + 1);
> txIds.push_back(partitionTxId);

yep, it's okay, but i am not sure that metaserver can handle this situation curretly which contain two same `PartitionTxId`, like:
```
partitionTxIds:
   - partitionId = 1
      txId = 2
   - partitionId = 1
      txId = 2
```
Please help us confirm this @SeanHai.

Reply:
@SeanHai had confirm it and i will improve it.